File: src/main/java/com/leijendary/spring/boot/template/model/LocalizedModel.java
package com.leijendary.spring.boot.template.model;

import static com.leijendary.spring.boot.template.util.RequestContext.getLanguage;
import static java.util.Comparator.comparingInt;
import static java.util.stream.Collectors.toList;
import static javax.persistence.FetchType.EAGER;

import java.util.HashSet;
import java.util.Set;

import javax.persistence.CollectionTable;
import javax.persistence.ElementCollection;
import javax.persistence.JoinColumn;
import javax.persistence.MappedSuperclass;

import lombok.Data;
import lombok.EqualsAndHashCode;
import lombok.ToString;

@Data
@EqualsAndHashCode(callSuper = true)
@MappedSuperclass
public abstract class LocalizedModel<T extends LocaleModel> extends UUIDModel {

    @ToString.Exclude
    @EqualsAndHashCode.Exclude
    @ElementCollection(fetch = EAGER)
    @CollectionTable(joinColumns = @JoinColumn(name = "id"))
    private Set<T> translations = new HashSet<>();

    public T getTranslation() {
        final var language = getLanguage();
        final var sorted = translations
                .stream()
                .sorted(comparingInt(LocaleModel::getOrdinal))
                .collect(toList());

        return sorted
                .stream()
                .filter(t -> t.getLanguage().equals(language))
                .findFirst()
                .orElse(sorted.iterator().next());
    }
}

File: src/main/kotlin/com/leijendary/spring/boot/template/core/model/LocalizedModel.kt
package com.leijendary.spring.boot.template.core.model

import com.leijendary.spring.boot.template.core.util.RequestContext.language
import javax.persistence.CollectionTable
import javax.persistence.ElementCollection
import javax.persistence.FetchType.EAGER
import javax.persistence.JoinColumn
import javax.persistence.MappedSuperclass

@MappedSuperclass
abstract class LocalizedModel<T : LocaleModel> : UUIDModel() {
    @ElementCollection(fetch = EAGER)
    @CollectionTable(joinColumns = [JoinColumn(name = "id")])
    val translations: Set<T> = HashSet()

    val translation: T
        get() {
            val language: String = language
            val sorted = translations.sortedBy { it.ordinal }

            return sorted.first { it.language == language }
        }
}